lib/auth/rotate.go

```diff
@@ -383,6 +389,92 @@ func (a *Server) autoRotate(ca types.CertAuthority) error {
 	return nil
 }
 
+type CertAuthorityMap = map[types.CertAuthType]types.CertAuthority
+
+// getAllCertificates returns all certificates authorities including private keys.
+func (a *Server) getAllCertificates(clusterName string) (CertAuthorityMap, error) {
+	certs := make(CertAuthorityMap)
+
+	for _, caType := range types.CertAuthTypes {
+		ca, err := a.Trust.GetCertAuthority(types.CertAuthID{
+			Type:       caType,
+			DomainName: clusterName,
+		}, true)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		certs[caType] = ca
+	}
+
+	return certs, nil
+}
+
+// findDuplicatedCertificates checks if any CA provided as caTypes has a duplicate in allCerts. List of all duplicates
+// is returned as a result. If no duplicates are found, then caTypes is returned in unmodified form.
+func (a *Server) findDuplicatedCertificates(caTypes []types.CertAuthType, allCerts CertAuthorityMap) []types.CertAuthType {
+	if len(caTypes) == len(allCerts) {
+		// requested rotation of all certs, no point of looking for duplicates
+		return caTypes
+	}
+
+	// create a set to prevent duplicates
+	rotateMap := make(map[types.CertAuthType]struct{})
+	certsThumbs := make(map[[32]byte][]types.CertAuthType)
+
+	for certType, allCert := range allCerts {
+		for _, cert := range allCert.GetActiveKeys().TLS {
+			s := sha256.Sum256(cert.Key)
```

Comment:
Not sure if checking cert iniquity based on cert.Key is the best option. Theoretical the same key can be used to generate severals CAs.

Reply:
In case when your CA is compromised I'd assume that you want to rotate everything that is signed with that key. Any suggestions how to find duplicates if we don't want to use the `Key`?